### ISSUE TITLE
Separate typechecking from codegenning

### ISSUE THREAD

**Issue by socumbersome** [rafalcieslak/CCoscope#21]
Refers to branch `typematcher`, and the issue itself refers to the comment https://github.com/rafalcieslak/CCoscope/pull/20#issuecomment-189512748 . 

Quick diagnosis -- we have mixed these two activities of our compiler which is probably not healthy :)

**Comment by rafalcieslak** [comment]
(continued discussion from the linked comment)

I believe we did not really have those stages separated in the first place, which is why I did not bother to mix them up a little bit right now. However, I expect that separating typechecking from codegenning should be very simple at this point! If I understand it correctly, then - using the terminology we spontaneously made up - the typechecking stage requires all nodes in the AST to determine their type (and recognize where conversions are needed), which is might be done just by asking them (recursivelly) what their `maintype` is. The result might be stored as a node member. This step, sometimes, requires a call to `Resolve()`, which is responsible for matching types and inserting conversions.

Thus, the way I see it, we probably want to split up the compilation into following stages:
1. Parse input and build AST (we currently treat it as a separate stage)
2. Ask all nodes to make up their mind about required conversions and their returned type. This is the only stage the typematcher should be used at. (Also, possibly restructure the AST applying optimizations.)
3. If the previous stage succeeded, use the information it produced to generate code. No need to `Resolve()` here (as we currently do), it's safe to assume all conversions etc. were inserted correctly.

Currently `maintype` is called only because it's occasionally needed by `codegens`. If we asked all nodes to determine their type before we start codegenning and then use cached data, then we naturally split the typechecking and code generation stages. It should require very little effort to restructure the compilation process this way.

**Comment by socumbersome** [comment]
That's correct, we started with these stages mixed but it wasn't as conspicuous as it is now.

Your proposal of three stages sounds good, but I still have problem with understanding the purpose of `Resolve`. I mean, because it exists just in order to call `TypeMatcher` at some point, then it **is** typechecking, so we should have either only `maintype`, or only `Resolve` -- there's no need to have both, I guess.

And just to make my point about separation -- for instance, `ReturnExprAST::Resolve` is supposed to be a typechecking phase, and yet it creates some code that `codegen` is responsible of creating (this happens when creating `MatchCandidateEntry` object).

I'm not entirely convinced that restructuring will require little effort, but I agree that it should be doable at least : )

**Comment by rafalcieslak** [comment]
There is indeed no need to keep both `maintype` and `Resolve`. They are currently separated, because at this point `codegen` needs to call `Resolve` (occasionally). Once we separate stages, it will no longer have to, so it'll be fine to merge `maintype` and `Resolve`.

`Resolve` never generates any code. However, it does prepare code-generating lambdas. This means that running `Resolve` does not generate code, it will only happen once you call the function that is stored within `match_result`. That function is indeed **prepared** during `Resolve` (as a lambda), but it is only run by `codegen`.

However, you may find it inelegant that the source code of `Resolve` contains parts that - though not run during `Resolve` - are responsible for generating code. Why is it currently this way? Because during typechecking `Resolve`, amongst others, figures out which variant to use for generating code (and defers that for later). For example, see `Resolve` for `BinaryExprAST` or `CallExprAST` (print variant). Depending on the result of typematcher decision, a different code has to be used for codegenning. This information has to be stored for later, because it will be required during `codegen`. Currently that information is stored in `match_result`, as an `std::function` called `assiciated_function`. This is a very flexible approach, as `Resolve` can insert virtually any result there. What are other options?
- We might use an `enum` instead of `std::function` to remember which variant has to be used during `codegen`. The `codegen` would be then a large `switch` over that enum, an each case would contain designated calls to llvm that insert requested instructions.
  - The problem with this approach is that only in some cases the result is indexed (like in case of printing a variable). Usually the result determines which variant of an operator to call, or which version of an overloaded function to use - and these options can not be not bound by any constant at compile time, as the user may define their custom overloads! (well, at some point they will be, hopefully)
- We might have each AST node use their own way to store information. `CallExprAST` would probably like to store the name of the function to call etc. However, the typematcher interface is generic for all nodes. In order to keep it that way, we would need to have a way of storing an arbitrary information, which means keeping a `void* user_data;` inside `MatchCandidateEntry`. Not only that becomes ugly and requires frequent casts, but it is also difficult to manage, as we would need to worry about the lifetime of that arbitrary data such pointer references.
- Another option I see would be to keep the lambdas in `Resolve`, but strip them to minimum so that they do not do any llvm calls, but instead pass the work to some other member functions of the node. For example, within `CallExprAST::Resolve` we might change:

<pre>
auto print_variant_int = MatchCandidateEntry{
    {ctx().getIntegerTy()},
    [this](std::vector<llvm::Value*> v){
        <b>return ctx().Builder.CreateCall(ctx().GetStdFunction("print_int"), v);</b>
    },
    ctx().getVoidTy()
};
</pre>

into

<pre>
auto print_variant_int = MatchCandidateEntry{
    {ctx().getIntegerTy()},
    [this](std::vector<llvm::Value*> v){
        <b>return this->GenerateStdLibCall("print_int", v);</b>
    },
    ctx().getVoidTy()
};
</pre>

On one hand, that would clean up the code a bit, especially in `ReturnExprAST::Resolve`, as it's main lambda would be a one-liner. On the other, we introduce various node-specific methods that generate code, and are used only by those lambdas - and the lambdas are only simple proxies for them (that, potentially, capture some values!).

But does that really solve the issue that `Resolve` seems to be generating code? Either way, it is only once you notice that these lambdas are never called by `Resolve` and only by `codegen`, that you realize that typechecking and codegenning is indeed separate (in this context). Wrapping the code generation in member methods instead of pure lambdas does not help in diminishing the illusion that these stages are intertwined (but, potentially, it makes the code slightly more readable).

Are there any other options for passing the information about which code generation variant shall be used when codegenning, that we should consider? Or is any of the above good enough for you?

**Comment by socumbersome** [comment]
From the dots you have described, I like third the most. And for me, node-specific methods are not so evil -- I'm not sure why would we want to avoid them (also, I didn't get the point about capturing values -- you can specify in a lambda what exactly should be captured, so there's no problem I guess?).

And yes, I understand that codegenning happen only in `codegen`, but the mere fact that there are parts responsible for codegenning in non-codegen context is kind of disturbing for me ; )

As to how I would see the refactoring -- I feel that there doesn't have to be much information available after typechecking. For instance, in:
-  `ReturnExprAST::Resolve()`, the codegenning code in a lambda should appear neither in `Resolve` nor in `maintype` I think, because the only thing needed later is `ConverterFunction`. So we could create `ConvertExprAST` and interleave "normal" AST nodes with such new nodes (these new storing appropriate `ConverterFunction`)
- `CallExprAST::Resolve()`, we could simply keep `Callee` as `llvm::Function*` alongside with `std::string` and typechecking would fill that field in. BTW, I think that keeping "real" AST node for function instead of just `std::string` would be better -- how do you think?

But all in all, I don't know whether I simply don't see some obvious obstacles to doing it this way. Anyway, I think that, if we have different visions :), we can work on refactoring on different branches and then discuss pros and cons of each approach and decide which will stay.

**Comment by rafalcieslak** [comment]
There is no problem with lambdas capturing variables - I just mean that the only purpose of such lambdas would be to bind a member method with a captured variable (which is totally fine).

I really like your ideas, especially inserting the conversions as intermediate AST nodes. I can't exactly imagine how to elegantly implement that (the structure of the tree would be modified while traversing it, unless maybe we are creating a new tree of a slightly different nature), so if you are confident about it, I would appreciate a simple draft for a proof-of-concept, or a detailed explanation.

It is probably clear that I tend to overgeneralize things. Quite likely we don't need to store arbitrary typeresolution result in nodes, I guess in practice we'll be just fine keeping the particular piece of data that is used by a given node.

I am unsure whether our visions are that much different, the way I understand it, these are the details we both agree on:
- The code from lambdas in `Resolve`s should be extracted to member functions.
- `codegen` should never call resolve, it should assume typechecking has been already done.
- `maintype` and `Resolve` should be merged into a single `Typecheck` method.
- Compilation should be performed by first calling (recursivelly) `Typecheck` on each node, and then (also recusrivelly) `codegenning` all of them.
- When possible `Typecheck` should insert `Conversion` nodes into the AST
- If any other result of the typechecking stage is required during codegenning, it should be stored in a (possibly non-public) member field of that particular node.

I feel like this is a sound plan.

**Comment by socumbersome** [comment]
Yes, this plan is something we should stick to. 
As to the insertion of `Conversion` nodes and how to implement this -- I think this will be the right moment to use `representative_` pointers at last. So, if object `A` of type `ExprAST` needs to have `Conversion` node above it, then we create new node `X := Conv(A)` and set `A.representative := X` and voila! I hope it will work ; ) (we will need some simple garbage collector in this immutable setting also).
Anyway, I'll probably don't have time until next weekend to start refactoring. But as we agree on most things and I hope the idea with representatives is clearer for you now, and if you have some spare time, feel free to start tinkering with it! Or you could focus on e.g. issue #23 as it looks as must-have.

**Comment by socumbersome** [comment]
Done. (https://github.com/rafalcieslak/CCoscope/commit/a34d6cd99d2fd7a67553d180f0be4e88de1d75f9)
